[patch 1/4] Network rework: Rewrite client/server packet handling & packet reading

nerzhul opened the pull request:
Hi,
packet handling isn't very good, it works but it need many test at each packet receive and it's done in only one function.

This pull request create a opcode table and a handler to handle each packet properly by using a opcodehandler mapping. This improve the network code performance and read.

ShadowNinja commented:
Seems like an improvement.  I don't know the network code well though (@sapier).

nerzhul commented:
Hi @ShadowNinja i fixed tab with spaces in https://github.com/nerzhul/minetest/commit/26edd4ecfb01762d5797ab72976e5ba3d70ca882

I will work on server side handling today, and if i get time i'll look at socket handling.

nerzhul commented:
@ShadowNinja in fact network code isn't handled well, the approach i propose is used since 6 year by mangos, a popular MMORPG emulator and is the best approach for a protocol handling.

I also think we must re-use unhandled opcodes instead of increasing them for future protocole usage before adding new.

This day i will work on server side. I also moved clientserver.h (which doesn't mean nothing) to networkprotocol.h, it's a better name.

Also, to have a better project view, i moved modified network stack sources to src/network/,

nerzhul commented:
Ok, i rewrite server side in a proper way. Server CPU load will decrease and some useless memory traces have been cleaned.

Now i have two thing to do:

- Rewrite packet handling by using an object which have the decoding methods which contains some useful informations (datasize & data at first)
- Add some helper for some duplicated code (like player getter into packet handlers)
- Add handler for depreciated functions (need packet handling rewrite)

nerzhul commented:
Hi,
i added new class, NetworkPacket, which contains the opcode, the data and some extracting methods (i'll add better methods with operators later).

Next step is to add helpers where it could be in handlers, and add those operators, and new handler for depreciated packets

sapier commented:
It's quite hard to review this code while it's split. please complete your work prior doing a pull request next time!

nerzhul commented:
Multiples commits permit to handle the changes, it's an agile method, then you can review the code whereas i'm writing it to improve it. It's a very very big update which permit to improve minetest network handling, and i really want it to work properly on networks :), and maybe become a minetest developper in the future, because i really love this game and i would help him to be powerful

My change isn't incompatible, because i tested a server with a new client packet handler and a client with new server packet handler, and new + new too, be careful ! no incompatible change here.

sapier commented:
I already removed the label ;-) Yet well still have to verify this as there's quite some history hidden within minetest code ... sadly. 
I'm positive about this change imho it's correct direction.
Please follow coding style use spaces around operators!

nerzhul commented:
Ok for coding style, i was disappointed because this code was very different in some things.

I'm working on an operator to extract and push datas to NetworkPacket, i think it's better to read packet with this method, used also by mangos.

https://github.com/mangoszero/server/blob/master/src/shared/ByteBuffer.h

ShadowNinja commented:
This is coming along nicely.

In case you don't know, the code style is written down here: http://dev.minetest.net/Code_style_guidelines
A lot of the current code doesn't follow it, but we're slowly converting things.

nerzhul commented:
Thanks @ShadowNinja 
for offset increment it's on the rails, i'm writing it :)

For peer_id i think it's a good idea, i don't know its use because i haven't looked where it's used, but it's a good idea. I'll do it after my rework on streaming interface

nerzhul commented:
Ok @ShadowNinja i added some of your suggestions, but i am tired. I do peer_id tomorrow and i'll try to remove many stringstream to use the >> operator i choose and add some missing helpers.

I also fixed NetworkPacket destructor

nerzhul commented:
Thanks for your tip @ShadowNinja i added some fixes.

rubenwardy commented:
Isn't this obsolete due to the ENet patch in progress?

nerzhul commented:
@rubenwardy link please ?
I haven't found a similar work

rubenwardy commented:
According to discussion IRC, @sapier is rebasing this to the latest version: https://github.com/sapier/minetest/tree/enet_support

Nevermind, anyway. It would be a 0.5 change, probably. This PR is still good.

nerzhul commented:
@rubenwardy right, ENet is the L4 approach, i'm on L7, that's good.

I notice some performance issues in minetest, i don't know if it's related to network but sometimes my server became at 100% CPU without anything.

I tested client loading on my 0.4.11 -stable server and client loading seems to be faster.

rubenwardy commented:
The networking does need a speed boost, both in bandwidth and latency. Freeminer uses ENet, they have a much larger data bandwidth because of it. They can transfer large sound files and textures quickly - faster loading.

How much more performance does this pull request give?
Does it break compatibility with older servers?

nerzhul commented:
@rubenwardy increase networking performance permit to have a better latency. For bandwidth, it's not my objective. Here it's a rewrite for improving maintenance and CPU cycle performance.

As i said in previous messages, if you look at my code:
- stable client can connect to this modified server
- modified client can connect to stable server
- modified client can connect to modified server.

Protocol hasn't change, but code was improved. Performance was gained into packet handling in one thing.
Before: look at opcode handling, it's a condition. Server and client do a IF/ELSE to match which packet it is and handle it into the ProcessData function
After: Packet opcode reference a function to handle it. It's faster, ~25-30 conditions are removed at each packet received.

And i also added a new class to handle packet like an object, not a buffer, which permit to read it easily (and in the future write it easily)

sfan5 commented on src/CMakeLists.txt:
```diff
@@ -472,6 +475,8 @@ set(minetest_SRCS
 	tile.cpp
 	wieldmesh.cpp
+	network/clientopcodes.cpp
+	network/toclientpacket.cpp
```
Wouldn't it make more sense to move these up to the top to the other files you added?

sfan5 commented on src/client.cpp:
```diff
@@ -1001,1008 +1001,1091 @@ void Client::Receive()
-	u8 ser_version = m_server_ser_ver;
+	if(pkt->getSize() >= 1+6+8+4)
+	{
+		// Get map seed
```
This comment is wrong.

nerzhul replied:
Fixed mainline

sfan5 commented on src/server.cpp:
```diff
@@ -1276,32 +1277,30 @@ PlayerSAO* Server::StageTwoClientInit(u16 peer_id)
 }
 
-void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+void Server::handleCommand_Depreciated(ToServerPacket* pkt)
```
should be `Deprecated`

nerzhul replied:
Fixed

sfan5 commented on src/server.h:
```diff
@@ -187,6 +188,35 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 	PlayerSAO* StageTwoClientInit(u16 peer_id);
+	
+	/*
+	 * Command Handler
```
Command Handler**s**

nerzhul replied:
Fixed in mainline (client.h + server.h), i'll wait more comments before commit

sfan5 commented on src/server.cpp:
```diff
@@ -1435,1301 +1437,1506 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
-						const ItemDefinition &hand = m_itemdef->get("");
-						const ToolCapabilities *tp = hand.tool_capabilities;
-						if(tp)
-							params = getDigParams(m_nodedef->ge
```
This is obvious, you don't need a comment for this.

nerzhul replied:
It's not my comment, but original comment :p, i remove it.

Zeno- commented:
Nice

nerzhul commented:
rebase is on the way

nerzhul commented:
rebase done. I'm working on memcpy improvement when creating NetworkPacket

Zeno- commented:
Once this is ready, I vote for it to be merged

nerzhul commented:
It's ready now. Future additions will be added in a future PR

sapier commented:
please don't mixup l7 and l4 changes.
nerzhul can you fix code style in this commit? There's a lot of code left where operators aren't surrounded by spaces.

nerzhul commented:
@sapier work on L4 and give me a buffer to read and set into NetworkPacket, then i handle it.
I think we must work together, maybe this week-end @sapier .

@sapier can you give me an example ?

sapier commented on src/client.cpp:
```diff
@@ -1001,1008 +1001,1097 @@ void Client::Receive()
 
-	if(command == TOCLIENT_INIT)
+	// Get player position
+	v3s16 playerpos_s16(0, BS*2+BS*20, 0);
```
missing spaces

sapier commented on src/client.cpp:
```diff
@@ -1001,1008 +1001,1097 @@ void Client::Receive()
-	if(command == TOCLIENT_INIT)
+	// Get player position
+	v3s16 playerpos_s16(0, BS*2+BS*20, 0);
+	if(pkt->getSize() >= 1+6)
```
missing spaces

sapier commented on src/client.cpp:
```diff
@@ -1001,1008 +1001,1097 @@ void Client::Receive()
-			return;
+		*pkt >> playerpos_s16;
+	}
+	v3f playerpos_f = intToFloat(playerpos_s16, BS) - v3f(0, BS/2, 0);
```
missing spaces

nerzhul commented:
@sapier i rework on old code coding style. It was fixed

sapier commented on src/client.cpp:
```diff
@@ -1001,1008 +1001,1077 @@ void Client::Receive()
+	float time_speed = 0;
 
-		assert(sector->getPos() == p2d);
+	if(pkt->getSize() >= 2+4) {
```
space missing

sapier commented on src/client.cpp:
```diff
@@ -1001,1008 +1001,1077 @@ void Client::Receive()
-		*/
-		addUpdateMeshTaskWithEdge(p, true);
+		if(m_time_of_day_set){
+			time_speed = (3600.0*24.0) * tod_diff_f / time_diff;
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1309,125 +1308,128 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
-		*/
+	u16 min_net_proto_version = 0;
+	if(pkt->getSize() >= 1 + PLAYERNAME_SIZE + PASSWORD_SIZE + 2)
+		min_net_proto_version = pkt->getU16(1+PLAYERNAME_SIZE+PASSWORD_SIZE);
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+	}
+	else
+	{
+		for(u32 i=0; i<PASSWORD_SIZE-1; i++)
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
-				DenyAccess(peer_id, narrow_to_wide(reason));
-				return;
-			}
+			given_password[i] = pkt->getChar(21+i);
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
-			}
+			given_password[i] = pkt->getChar(21+i);
 		}
+		given_password[PASSWORD_SIZE-1] = 0;
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+		Answer with a TOCLIENT_INIT
+	*/
+	{
+		SharedBuffer<u8> reply(2+1+6+8+4);
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+		writeU16(&reply[0], TOCLIENT_INIT);
+		writeU8(&reply[2], deployed);
+		//send dummy pos for legacy reasons only
+		writeV3S16(&reply[2+1], floatToInt(v3f(0,0,0), BS));
```
space missing following lines too

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+
+	RemoteClient *client = getClient(pkt->getPeerId());
+
+	for(u16 i=0; i<count; i++)
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+
+	for(u16 i=0; i<count; i++)
+	{
+		if((s16)pkt->getSize() < 1+(i+1)*6)
```
space missing

sapier commented on src/server.cpp:
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+
+void Server::handleCommand_PlayerPos(ToServerPacket* pkt)
+{
+	if(pkt->getSize() < 12+12+4+4)
```
space missing ... stopping to check here please check following code yourself

sapier commented:
seems to be quite fine. As this is a change in critical code area I requested VanessaE to run it on her servers. If she doesn't complain about errors till January 23rd I'd suggest merging it.

nerzhul commented:
@sapier ok thanks, i stay on this topic to listen to any change to do.
I'm working on packet writing and send cleanup in a second PR.

Zeno- commented:
@sapier This patch is also applied and running on mtzeno.com (30000)

nerzhul commented:
Rebased to current master

Zeno- commented:
This has been running on my server for almost 2 weeks and I haven't noticed any issues and haven't received any reports. The code looks fine.

+1 for merging as soon as feature freeze is over

nerzhul commented:
Hi Zeno, it was rebased to prevent conflict with my recent security fix

Zeno- commented:
Rebase is good ;)

Zeno- commented:
commit a704c04f00bfea4b77550169fa08105c2ee0dfd0
